Strapi MCP server for content management (#3194)
strapi/documentation: docusaurus/docs/cms/features/strapi-mcp-server.md

---
title: MCP server
displayed_sidebar: cmsSidebar
description: Expose content management tools to AI clients through a built-in MCP server.
tags:
  - features
  - ai
  - MCP
  - content management
toc_max_heading_level: 4
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# MCP server

> <VersionBadge version="5.48.0"/>

Reviewer: Next milestone is 5.47
```suggestion
<VersionBadge version="5.47.0"/>
```

  <IdentityCardItem icon="desktop" title="Environment">Available in both Development & Production environment</IdentityCardItem>
</IdentityCard>

## Configuration

Before first use, the Strapi MCP server must be:
- enabled through the server configuration file and authenticated with Admin tokens created in the admin panel
- connected to your AI client.

### Strapi code-based configuration

Enable the MCP server by adding the `mcp` object to the server configuration file:

<Tabs groupId="js-ts">
<TabItem value="javascript" label="JavaScript">

```js title="config/server.js"
module.exports = ({ env }) => ({
  host: env('HOST', '0.0.0.0'),
  port: env.int('PORT', 1337),
  app: {
    keys: env.array('APP_KEYS'),
  },
  // highlight-start
  mcp: {
    enabled: true,
  },
  // highlight-end
});
```

</TabItem>
<TabItem value="typescript" label="TypeScript">

```ts title="config/server.ts"
import type { Core } from '@strapi/strapi';

const config = ({ env }: Core.Config.Shared.ConfigParams): Core.Config.Server => ({
  host: env('HOST', '0.0.0.0'),
  port: env.int('PORT', 1337),
  app: {
    keys: env.array('APP_KEYS'),
  },

>   mcp: {
>     enabled: true,
>   },

Reviewer: ??
```suggestion
  // highlight-start
  mcp: {
    enabled: true,
  },
  // highlight-end
```

});
export default config;
```

</TabItem>
</Tabs>

Once the setting is in place, restart Strapi. The MCP endpoint becomes available at `/mcp` on your Strapi server (e.g., `http://localhost:1337/mcp`).

#### Advanced options

The following optional keys can be added to the `mcp` configuration object:

| Option | Type | Default | Description |
|--------|------|---------|-------------|
| `enabled` | Boolean | `false` | Enable or disable the MCP server. |
| `connectTimeoutMs` | Number | `5000` | Maximum time (in ms) for the internal MCP transport to connect before the request is aborted. |
| `requestTimeoutMs` | Number | `60000` | Maximum time (in ms) for a single MCP request to complete before it times out. |

```js title="config/server.js"
mcp: {
  enabled: true,
  connectTimeoutMs: 10000, // 10 seconds
  requestTimeoutMs: 120000, // 2 minutes
},
```

:::note
The timeout options are read at runtime and are not yet included in Strapi's `Server` TypeScript configuration type. If you use TypeScript, the basic `mcp: { enabled: true }` object is fully typed, but adding timeout keys may require a type assertion.

> :::

Reviewer: Fixed in the implementation

```suggestion
```

### Strapi admin panel configuration

The MCP server authenticates requests using Admin tokens. Each MCP session is scoped to the permissions of the token used to connect:

1. Create a new Admin token (see [Creating an Admin token](/cms/features/admin-tokens#creating-a-new-admin-token) on the Admin tokens feature page).
2. Copy the token value. You will need it when configuring the AI client.

The token's permissions determine which MCP tools are exposed to the AI client. For instance, if the token only grants `read` on an `Article` content-type, the AI client will only see listing and reading tools for articles.

### AI client configuration

Once you have enabled the MCP server through the server configuration file and created an Admin token in the admin panel, connect your AI client to the Strapi MCP server.

:::note

> `localhost:1337/` is used in configuration examples on this page. If your Strapi server is hosted on another URL or port, please update the code accordingly.

Reviewer: ```suggestion
`http://localhost:1337/` is used in configuration examples on this page. If your Strapi server is hosted on another URL or port, please update the code accordingly.
```

| Tool | Action | Permission required | Description |
|------|--------|-------------------|-------------|
| `list` | Read | `read` | List entries with pagination, sorting, and filtering |
| `get` | Read | `read` | Get a single entry by document ID |
| `create` | Create | `create` | Create a new entry (created as draft if Draft & Publish is enabled) |
| `update` | Update | `update` | Update an existing entry by document ID |
| `delete` | Delete | `delete` | Delete an entry by document ID |
| `publish` | Publish | `publish` | Publish a draft entry |
| `unpublish` | Unpublish | `publish` | Unpublish a published entry |
| `discard_draft` | Discard draft | `publish` | Discard draft changes and revert to the published version |

The last 3 tools (publish, unpublish, discard_draft) are only generated for content types that have [Draft & Publish](/cms/features/draft-and-publish) enabled.

Tool names follow the pattern `cm_<namespace>_<model>_<action>`. For example, an `Article` content type with UID `api::article.article` generates tools named `cm_api_article_list`, `cm_api_article_get`, `cm_api_article_create`, etc.

> Single types follow a similar pattern but include `_single_` in the name (e.g., `cm_api_homepage_single_get`). Because single types represent a unique entry, they do not have a `list` tool and generate up to 7 tools instead of 8.

Reviewer: Tool name will change, not sure this has a value to be documented here.
```suggestion
```

#### Built-in utility tools

In addition to content-type tools, Strapi registers the following built-in tools:

| Tool | Availability | Description |
|------|-------------|-------------|
| `log` | Development mode only | Logs a message to the Strapi server console at a specified level (`info`, `warn`, `error`, `http`, `log`). Useful for debugging MCP interactions. |

> Built-in utility tools are only available in development mode (when `autoReload` is enabled) and do not require specific content-type permissions.

Reviewer: ```suggestion
Built-in utility tools are only available in development mode (when `autoReload` is enabled) and do not require specific admin permissions.
```

### Input schemas

Each tool has an input schema derived from your content type's attributes. The schema is dynamically generated per session based on the token's permissions:

- **Field-level permissions**: If the token restricts access to certain fields, those fields are excluded from both input and output schemas. For instance, a token that can only read `title` and `slug` on an `Article` type will only see those 2 fields in sort, filter, and output schemas.
- **Locale-level permissions**: If the [Internationalization (i18n)](/cms/features/internationalization) feature is enabled and the token restricts access to certain locales, the `locale` parameter only accepts the permitted locales for each action. Locale restrictions are evaluated per action: a token might allow reading in `en` and `fr` but only creating in `en`.
- **Write operations**: The `data` object in `create` and `update` tools only includes fields the token has permission to write. Unknown fields are rejected (strict validation). System-managed fields (`id`, `documentId`, `createdAt`, `updatedAt`, `publishedAt`, `createdBy`, `updatedBy`) are excluded automatically. Private fields are also excluded.

> - **Attribute type mapping**: Field schemas carry constraints from your content-type definition, such as `required`, `minLength`, `maxLength`, `min`, `max`, and `enum` values. Enumeration fields expose their allowed values in the schema so the AI client can pick valid options.

Reviewer: Feels really redundant with following `### Permission boundaries`
```suggestion
```

| Object | `{ "title": "asc" }` |
| Array of objects | `[{ "title": "asc" }, { "createdAt": "desc" }]` |

Sort field names are constrained to the content type's scalar attributes (strings, numbers, booleans, dates, enumerations). Relation, component, dynamic zone, media, and JSON fields cannot be sorted on.

#### Filtering

The `list` tool accepts a `filters` parameter using Strapi's filter syntax:

- **Field operators**: `$eq`, `$ne`, `$in`, `$notIn`, `$lt`, `$lte`, `$gt`, `$gte`, `$between`, `$contains`, `$notContains`, `$startsWith`, `$endsWith`, `$null`, `$notNull`, and their case-insensitive variants (`$eqi`, `$nei`, `$containsi`, `$notContainsi`, `$startsWithi`, `$endsWithi`).
- **Logical operators**: `$and`, `$or` (accept arrays of filter objects), `$not` (wraps a single filter object).
- **Implicit equality**: Passing a value directly (e.g., `{ "title": "Hello" }`) is equivalent to `{ "title": { "$eq": "Hello" } }`.

Like sort fields, filter fields are constrained to scalar attributes only.

#### Pagination

The `list` tool also accepts `page` (1-indexed, default: 1) and `pageSize` (default: 25, max: 100) parameters.

#### Rich text (blocks)

> Rich text fields using the [Blocks editor](/cms/features/content-type-builder#rich-text-blocks) are represented with a structured schema in tool definitions. The schema covers all block types: paragraphs, headings (levels 1–6), quotes, code blocks (with optional language), ordered and unordered lists (including nested lists), images, and inline nodes (text with formatting marks, links).

Reviewer: Not sure this is relevant to document
```suggestion
```

GET and DELETE HTTP methods on the `/mcp` endpoint return a `405 Method Not Allowed` JSON-RPC error, as the MCP server only accepts POST requests.

### Known limitations

The MCP server has the following limitations:

- **Components**: Component fields are passed as untyped (`any`) in tool schemas. The AI client can read and write component data, but the schema does not describe the component's internal structure.
- **Dynamic zones**: Dynamic zone fields are passed as untyped arrays in tool schemas.
- **Relations**: Relation fields accept document IDs (e.g., `"z7v8zma53x01r6oceimv922b"`) as input. The advanced `connect`/`disconnect` relation syntax is not yet supported in MCP tool schemas.
- **Nested population parameters**: The `list` and `get` tools do not support nested population parameters for relations.
- **Media upload**: Media fields accept existing media asset references but the MCP server cannot upload new files. Use Strapi's media library or upload API to add files first, then reference them in MCP tool calls.

> - **Custom fields**: Custom fields registered via plugins are mapped to their underlying Strapi type. If the custom field registry is not populated when MCP tools are registered, the custom field falls back to an `unknown` type.

Reviewer: This might/should change before release

Author: Noted, Nico, thanks!
If it's updated while I'm off, feel free to create suggestions in GH comments and commit them 😊

| "Update article abc123, change the title to 'Hello Strapi'." | Updates the title, other fields untouched |
| "Publish article abc123." | Changes the entry status to published |
| "Delete article abc123." | Removes the entry |
| "Create an article in French with the title 'Bonjour le monde'." | Creates a draft article with `locale` set to `fr` |

> #### Internationalization (i18n)

Author: @nclsndr Added this part to try to handle [this QA issue](https://www.notion.so/strapi/i18n-Isn-t-aware-of-locales-on-content-type-3658f35980748004b8d7c8aa8b800570?d=2a98f3598074820ebab303116df75e76)
WDYT?

Reviewer: Good idea @pwizla 👌
